devito/types/dimension.py

```diff
@@ -711,8 +711,15 @@ class ConditionalDimension(DerivedDimension):
     def __init_finalize__(self, name, parent, factor=None, condition=None,
                           indirect=False):
         super().__init_finalize__(name, parent)
+
+        # TODO: Avoid local import of lower_exprs
+        from devito.ir.equations.algorithms import lower_exprs
+
+        expressions = []
+        expressions.append(condition)
+
         self._factor = factor
-        self._condition = condition
+        self._condition = lower_exprs(expressions).pop()
```

Comment:
wait so this is done at `__init__()`. that' a bit odd isn't it?
So I assume this is because Dimension are never lowered. It used to be indexified somewhere though jsut without the shift, where was that happening?

Reply:
no, it's fine to do it at `__init__`. The issue is that `condition` is an actual expression, possible with lots of indexed, and this is the only place we can lower it without losing sanity. Because, if we instead lowered it later on, during Operator compilation, then to avoid changing state we would have to go through an incredible mess:
* locating all CDs
* getting their conditions and explicitly lowering them
* create new CDs with the new lowered conditions
* replace the old CDs with new CDs

and even then, this would lead to all sort of headaches. Not to mention when one uses the CD directly inside a Function...

so this is the only place we can perform this lowering, and since lowering is a local thing, I don't think there's anything bad with it

Reply:
Makes sense. Not part of the repr so shouldn't be confusing neither.

Reply:
So this should be `self._condition = expression` and lowering moved to algorithms no?
https://github.com/devitocodes/devito/blob/6aa03bb38b74224f742beb1dbe8df814b45941a0/devito/ir/clusters/algorithms.py#L179